lib/jxl/encode.cc

```diff
@@ -489,23 +489,34 @@ JxlEncoderStatus JxlEncoderSetBasicInfo(JxlEncoder* enc,
     return JXL_ENC_NOT_SUPPORTED;
   }
 
-  switch (info->alpha_bits) {
-    case 0:
-      break;
-    case 32:
-    case 16:
-      enc->metadata.m.SetAlphaBits(16);
-      break;
-    case 8:
-      enc->metadata.m.SetAlphaBits(info->alpha_bits);
-      break;
-    default:
-      return JXL_ENC_ERROR;
-  }
   // The number of extra channels includes the alpha channel, so for example and
   // RGBA with no other extra channels, has exactly num_extra_channels == 1
   enc->metadata.m.num_extra_channels = info->num_extra_channels;
   enc->metadata.m.extra_channel_info.resize(enc->metadata.m.num_extra_channels);
+  if (info->num_extra_channels == 0 && info->alpha_bits) {
+    return JXL_API_ERROR(
+        "when alpha_bits is non-zero, the number of channels must be at least "
+        "1");
+  }
+  // If the user provides non-zero alpha_bits, we make the channel info at index
+  // zero the appropriate alpha channel.
+  if (info->alpha_bits) {
+    JxlExtraChannelInfo channel_info;
+    JxlEncoderInitExtraChannelInfo(JXL_CHANNEL_ALPHA, &channel_info);
+    switch (info->alpha_bits) {
+      case 32:
+      case 16:
+        channel_info.bits_per_sample = 16;
+        break;
+      case 8:
+        channel_info.bits_per_sample = info->alpha_bits;
```

Comment:
Actually now that I'm thinking about it: why are we restricting it to be 8 or 16?

There should be no issue having an alpha channel that is, say, 10-bit.

How the pixels are passed to the encoder or from the decoder is a different thing, but that's not what this part of the API is about (we have pixelformat for that, and yes, there it has to be 8-bit, 16-bit, or I guess float).

So I think we can just set `channel_info.bits_per_sample = info->alpha_bits`.
We should also set `channel_info.exponent_bits_per_sample = info->alpha_exp_bits`, for the floating-point case.

SetExtraChannelInfo should take care of invalid values then, but you'll have to get its return status.

Reply:
Ok, I agree that there should be no problem having e.g. 10-bit alpha. 
(And right now this is already possible by manually defining the extra channel. 
I'm not sure what that weird switch-case was meant for. 
Setting `bits_per_sample` here to unreasonable values, eventually results in failures with meaningful error messages, coming from `VisitFields` in `image_metadata.cc`.
Should I add more checks in `setExtraChannelInfo`? If yes, which kind?

Reply:
The conditions on bits_per_sample and exponent_bits_per_sample are identical between the color channels and the extra channels, so it would be good to use the same check when setting it.

Reply:
I doubled the check from the beginning of `JxlEncoderSetBasicInfo` to `JxlEncoderSetExtraChannelInfo`, so using the same conditions for color channels and extra channels now, please check if this looks reasonable.